### PR TITLE
[jit] Add Python language reference docs

### PR DESCRIPTION
**Pull request by driazati**
This exposes our audit of https://docs.python.org/3/reference/ with descriptions for each line item.

To generate the `.rst` from the Quip:

```bash
pip install m2r
m2r jit_language_reference.md
```

https://driazati.github.io/pytorch_doc_previews/30686/jit.html#python-functions-and-modules

Differential Revision: [D19219587](https://our.internmc.facebook.com/intern/diff/19219587/)

### CODE REVIEW

**Comment by eellison** [comment]
can you setup a github repo so i can view the docs in html ?

**Comment by driazati** [comment]
It's already linked in the PR's description

File: docs/source/jit_language_reference.rst [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+     - Not supported: **=, <<=, >>=, %=, ^=, @=, &=, //=, % for strings
+   * - `3. Data model <https://docs.python.org/3/reference/datamodel.html#>`_
+     - Not Relevant
+     -
+   * - `3.1. Objects, values and types <https://docs.python.org/3/reference/datamodel.html#objects-values-and-types>`_
+     - Not Relevant
+     - id() not supported, type() upcalls to Python
+   * - `3.2. The standard type hierarchy <https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy>`_
+     - Partially Supported
+     - op == with None works in Python, not supported: NotImplemented, Ellipsis, numbers.Complex, bytes, byte arrays, sets, frozen sets, generators, coroutines, async generators, modules, I/O objects, internal objects, slice objects ( though slicing is supported), classmethod
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I would remove `op == with None works in Python` and just list the not supported section.

File: docs/source/jit_language_reference.rst [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+   * - `3.3. Special method names <https://docs.python.org/3/reference/datamodel.html#special-method-names>`_
+     - Supported
+     -
+   * - `3.3.1. Basic customization <https://docs.python.org/3/reference/datamodel.html#basic-customization>`_
+     - Partially Supported
+     - Not supported: **new**\ , **del**\ , **bytes**\ , **format**\ , **hash**\ ,
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: why bold these methods but not bold the unsupported type hierarchy list ?

File: docs/source/jit_language_reference.rst [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+   * - `3.3.2.2. Implementing Descriptors <https://docs.python.org/3/reference/datamodel.html#implementing-descriptors>`_
+     - Not Supported
+     -
+   * - `3.3.2.3. Invoking Descriptors <https://docs.python.org/3/reference/datamodel.html#invoking-descriptors>`_
+     - Not Supported
+     -
+   * - `3.3.2.4. **slots** <https://docs.python.org/3/reference/datamodel.html#slots>`_
+     - Not Supported
+     -
+   * - `3.3.2.4.1. Notes on using **_slots**\ * <https://docs.python.org/3/reference/datamodel.html#notes-on-using-slots>`_
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Probably don't need both `3.3.2.4.` and `3.3.2.4.1.` since both are just `Not Supported`

File: docs/source/jit_language_reference.rst [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `3.3.3. Customizing class creation <https://docs.python.org/3/reference/datamodel.html#customizing-class-creation>`_
+     - Not Supported
+     -
+   * - `3.3.3.1. Metaclasses <https://docs.python.org/3/reference/datamodel.html#metaclasses>`_
+     - Not Supported
+     -
+   * - `3.3.3.2. Resolving MRO entries <https://docs.python.org/3/reference/datamodel.html#resolving-mro-entries>`_
+     - Not Supported
+     - No inheritance
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: No inheritance is a little misleading, because we do have some amount of support for inheritance within modules

File: docs/source/jit_language_reference.rst [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+   * - `3.3.3.7. Uses for metaclasses <https://docs.python.org/3/reference/datamodel.html#uses-for-metaclasses>`_
+     - Not relevant
+     -
+   * - `3.3.4. Customizing instance and subclass checks <https://docs.python.org/3/reference/datamodel.html#customizing-instance-and-subclass-checks>`_
+     - Not Supported
+     -
+   * - `3.3.5. Emulating generic types <https://docs.python.org/3/reference/datamodel.html#emulating-generic-types>`_
+     - Not Supported
+     -
+   * - `3.3.6. Emulating callable objects <https://docs.python.org/3/reference/datamodel.html#emulating-callable-objects>`_
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: This is supported no: `__call__` ?

File: docs/source/jit_language_reference.rst [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+     - Supported
+     -
+   * - `3.3.7. Emulating container types <https://docs.python.org/3/reference/datamodel.html#emulating-container-types>`_
+     - Partially Supported
+     - some magic methods not supported (e.g. **iter**\ )
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: This wouldn't be that hard to add...

File: docs/source/jit_language_reference.rst [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `3.4.3. Asynchronous Iterators <https://docs.python.org/3/reference/datamodel.html#asynchronous-iterators>`_
+     - Not Supported
+     -
+   * - `3.4.4. Asynchronous Context Managers <https://docs.python.org/3/reference/datamodel.html#asynchronous-context-managers>`_
+     - Not Supported
+     -
+   * - `4. Execution model <https://docs.python.org/3/reference/executionmodel.html#>`_
+     - Not Relevant
+     - Not directly relevant but this could use some explanation on how TorchScript differs from Python
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: `Not directly relevant but this could use some explanation on how TorchScript differs from Python` not sure what this means

File: docs/source/jit_language_reference.rst [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+   * - `4.1. Structure of a program <https://docs.python.org/3/reference/executionmodel.html#structure-of-a-program>`_
+     - Not Relevant
+     -
+   * - `4.2. Naming and binding <https://docs.python.org/3/reference/executionmodel.html#naming-and-binding>`_
+     - Not Relevant
+     - Names are entriely bound at compile time in TorchScript
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: `/s/entriely/entirely/`

File: docs/source/jit_language_reference.rst [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+   * - `4.2.1. Binding of names <https://docs.python.org/3/reference/executionmodel.html#binding-of-names>`_
+     - Not Relevant
+     - See ``global`` and ``nonlocal`` statements
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: What does this means `See ``global`` and ``nonlocal`` statements` ?  if it's a link to another section can u add the ref

File: docs/source/jit_language_reference.rst [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+   * - `4.2.2. Resolution of names <https://docs.python.org/3/reference/executionmodel.html#resolution-of-names>`_
+     - Not Relevant
+     - See ``global`` and ``nonlocal`` statements
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Same comment

File: docs/source/jit_language_reference.rst [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+   * - `4.2.3. Builtins and restricted execution <https://docs.python.org/3/reference/executionmodel.html#builtins-and-restricted-execution>`_
+     - Not Relevant
+     -
+   * - `4.2.4. Interaction with dynamic features <https://docs.python.org/3/reference/executionmodel.html#interaction-with-dynamic-features>`_
+     - Not Supported
+     - Python values cannot be captured so this is not an issue
+   * - `4.3. Exceptions <https://docs.python.org/3/reference/executionmodel.html#exceptions>`_
+     - Partially Supported
+     - See ``try`` and ``raise`` statement
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Same comment: could you add ref, not really sure what this means without it

File: docs/source/jit_language_reference.rst [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+   * - `5. The import system <https://docs.python.org/3/reference/import.html>`_
+     - Not Relevant
+     -
+   * - `6. Expressions <https://docs.python.org/3/reference/expressions.html#>`_
+     - Not Relevant
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Link here? https://driazati.github.io/pytorch_doc_previews/30686/jit.html#expressions

File: docs/source/jit_language_reference.rst [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `6.1. Arithmetic conversions <https://docs.python.org/3/reference/expressions.html#arithmetic-conversions>`_
+     - Partially Supported
+     - Complex number coercion doesn't work, but (float + int) -> float does work
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I think you can say supported, we've already said we don't support complex numbers. Otherwise you can just say only complex numbers not supported or something.

File: docs/source/jit_language_reference.rst [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+   * - `6.2. Atoms <https://docs.python.org/3/reference/expressions.html#atoms>`_
+     - Not Relevant
+     -
+   * - `6.2.1. Identifiers (Names) <https://docs.python.org/3/reference/expressions.html#atom-identifiers>`_
+     - Supported
+     -
+   * - `6.2.2. Literals <https://docs.python.org/3/reference/expressions.html#literals>`_
+     - Partially Supported
+     - bytesliteral, imagnumber not supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: `/s/imagnumber/imagenumber/`. also, same as above, consistency when listing things and whether they're bolded or not please.

File: docs/source/jit_language_reference.rst [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+     - Not supported: comprehension ifs, async iterators
+   * - `6.2.5. List displays <https://docs.python.org/3/reference/expressions.html#list-displays>`_
+     - Supported
+     -
+   * - `6.2.6. Set displays <https://docs.python.org/3/reference/expressions.html#set-displays>`_
+     - Not Supported
+     -
+   * - `6.2.7. Dictionary displays <https://docs.python.org/3/reference/expressions.html#dictionary-displays>`_
+     - Supported
+     - dict() constructor with kwargs doesn't work, dict comprehensions, dictionary unpacking, limited key types (hashable TorchScript classes??)
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: maybe instead: `not supported: dict() constructor with kwargs, dict comprehensions, ...`. Maybe rephrase or remove `limited key types (hashable TorchScript classes??)`

File: docs/source/jit_language_reference.rst [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `6.2.9.4. Asynchronous generator-iterator methods <https://docs.python.org/3/reference/expressions.html#asynchronous-generator-iterator-methods>`_
+     - Not Supported
+     -
+   * - `6.3. Primaries <https://docs.python.org/3/reference/expressions.html#primaries>`_
+     - Supported
+     -
+   * - `6.3.1. Attribute references <https://docs.python.org/3/reference/expressions.html#attribute-references>`_
+     - Partially Supported
+     - Attribute references on primtive types in Python (int) fail in TorchScript
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: `/s/primtive/primitive`. also i don't know what attribute references on primitive types in Python (int) fail maenas

File: docs/source/jit_language_reference.rst [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+   * - `6.3.2. Subscriptions <https://docs.python.org/3/reference/expressions.html#subscriptions>`_
+     - Supported
+     -
+   * - `6.3.3. Slicings <https://docs.python.org/3/reference/expressions.html#slicings>`_
+     - Partially Supported
+     - Tuple slicing with stride doesn't work
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Maybe: consistency with `Not supported: ...` vs `x doesn't work`

File: docs/source/jit_language_reference.rst [hunk 18]
```diff
@@ -0,0 +1,438 @@
...
+   * - `6.3.4. Calls <https://docs.python.org/3/reference/expressions.html#calls>`_
+     - Partially Supported
+     - args unpack / kwargs unpack doesn't work
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: same comment ^^

File: docs/source/jit_language_reference.rst [hunk 19]
```diff
@@ -0,0 +1,438 @@
...
+   * - `6.4. Await expression <https://docs.python.org/3/reference/expressions.html#await-expression>`_
+     - Not Supported
+     -
+   * - `6.5. The power operator <https://docs.python.org/3/reference/expressions.html#the-power-operator>`_
+     - Supported
+     -
+   * - `6.6. Unary arithmetic and bitwise operations <https://docs.python.org/3/reference/expressions.html#unary-arithmetic-and-bitwise-operations>`_
+     - Partially Supported
+     - negate implemented only for tensors
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: What does this mean ? `negate implemented only for tensors` ?

File: docs/source/jit_language_reference.rst [hunk 20]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `6.9. Binary bitwise operations <https://docs.python.org/3/reference/expressions.html#binary-bitwise-operations>`_
+     - Supported
+     -
+   * - `6.10. Comparisons <https://docs.python.org/3/reference/expressions.html#comparisons>`_
+     - Supported
+     -
+   * - `6.10.1. Value comparisons <https://docs.python.org/3/reference/expressions.html#value-comparisons>`_
+     - Partially Supported
+     - Maybe None == None should throw an error since the ref says not to do it, list/dict equality
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I would remove: `Maybe None == None should throw an error since the ref says not to do it`. What do you mean by list/dict equality?

File: docs/source/jit_language_reference.rst [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `6.5. The power operator <https://docs.python.org/3/reference/expressions.html#the-power-operator>`_
+     - Supported
+     -
+   * - `6.6. Unary arithmetic and bitwise operations <https://docs.python.org/3/reference/expressions.html#unary-arithmetic-and-bitwise-operations>`_
+     - Partially Supported
+     - negate implemented only for tensors
+   * - `6.7. Binary arithmetic operations <https://docs.python.org/3/reference/expressions.html#binary-arithmetic-operations>`_
+     - Partially Supported
+     - See delimiters
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Link to delimiters if you can

File: docs/source/jit_language_reference.rst [hunk 22]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `6.10. Comparisons <https://docs.python.org/3/reference/expressions.html#comparisons>`_
+     - Supported
+     -
+   * - `6.10.1. Value comparisons <https://docs.python.org/3/reference/expressions.html#value-comparisons>`_
+     - Partially Supported
+     - Maybe None == None should throw an error since the ref says not to do it, list/dict equality
+   * - `6.10.2. Membership test operations <https://docs.python.org/3/reference/expressions.html#membership-test-operations>`_
+     - Partially Supported
+     - **iter** lookup not supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: How do you mean `iter` not supported?

File: docs/source/jit_language_reference.rst [hunk 23]
```diff
@@ -0,0 +1,438 @@
...
+   * - `6.10.3. Identity comparisons <https://docs.python.org/3/reference/expressions.html#is-not>`_
+     - Supported
+     -
+   * - `6.11. Boolean operations <https://docs.python.org/3/reference/expressions.html#boolean-operations>`_
+     - Supported
+     -
+   * - `6.12. Conditional expressions <https://docs.python.org/3/reference/expressions.html#conditional-expressions>`_
+     - Partially Supported
+     - if + lambda expr not supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I think we can just say Conditional expressions is supported

File: docs/source/jit_language_reference.rst [hunk 24]
```diff
@@ -0,0 +1,438 @@
...
+     - Partially Supported
+     - iterable unpacking not supported
+   * - `6.15. Evaluation order <https://docs.python.org/3/reference/expressions.html#evaluation-order>`_
+     - Supported
+     -
+   * - `6.16. Operator precedence <https://docs.python.org/3/reference/expressions.html#operator-precedence>`_
+     - Supported
+     -
+   * - `7. Simple statements <https://docs.python.org/3/reference/simple_stmts.html#>`_
+     - Irrelevant
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I would say supported

File: docs/source/jit_language_reference.rst [hunk 25]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `7.1. Expression statements <https://docs.python.org/3/reference/simple_stmts.html#expression-statements>`_
+     - Supported
+     -
+   * - `7.2. Assignment statements <https://docs.python.org/3/reference/simple_stmts.html#assignment-statements>`_
+     - Partially Supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: How do we not support this, i think we do?

File: docs/source/jit_language_reference.rst [hunk 26]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `7.2.1. Augmented assignment statements <https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements>`_
+     - Partially Supported
+     - A bunch of these don't work, see "Delimiters"
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Link: to ref

File: docs/source/jit_language_reference.rst [hunk 27]
```diff
@@ -0,0 +1,438 @@
...
+   * - `7.2.2. Annotated assignment statements <https://docs.python.org/3/reference/simple_stmts.html#annotated-assignment-statements>`_
+     - Supported
+     -
+   * - `7.3. The assert statement <https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement>`_
+     - Partially Supported
+     - Message is not customized (it gets desugared into a ``RaiseException``\ )
+   * - `7.4. The pass statement <https://docs.python.org/3/reference/simple_stmts.html#the-pass-statement>`_
+     - Partially Supported
+     - String parser error with 1 line functions
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: string parser not user facing, i would remove this

File: docs/source/jit_language_reference.rst [hunk 28]
```diff
@@ -0,0 +1,438 @@
...
+   * - `7.5. The del statement <https://docs.python.org/3/reference/simple_stmts.html#the-del-statement>`_
+     - Not Supported
+     -
+   * - `7.6. The return statement <https://docs.python.org/3/reference/simple_stmts.html#the-return-statement>`_
+     - Supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I think we can just say supported. i dont know how try ... finally is relevant here

File: docs/source/jit_language_reference.rst [hunk 29]
```diff
@@ -0,0 +1,438 @@
...
+     - Some other features of returning (e.g. behavior with try..finally) are unsupported
+   * - `7.7. The yield statement <https://docs.python.org/3/reference/simple_stmts.html#the-yield-statement>`_
+     - Not Supported
+     -
+   * - `7.8. The raise statement <https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement>`_
+     - Partially Supported
+     - Message is not customized, ... from statements silently do nothing
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: what does `...` from statements silently do nothing  mean ?

File: docs/source/jit_language_reference.rst [hunk 30]
```diff
@@ -0,0 +1,438 @@
...
+   * - `7.9. The break statement <https://docs.python.org/3/reference/simple_stmts.html#the-break-statement>`_
+     - Supported
+     - Some other features of returning (e.g. behavior with try..finally) are unsupported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I think you can just say supported. there is a separate section for try except

File: docs/source/jit_language_reference.rst [hunk 31]
```diff
@@ -0,0 +1,438 @@
...
+   * - `7.10. The continue statement <https://docs.python.org/3/reference/simple_stmts.html#the-continue-statement>`_
+     - Supported
+     - Some other features of returning (e.g. behavior with try..finally) are unsupported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Same: i think you can just say supported

File: docs/source/jit_language_reference.rst [hunk 32]
```diff
@@ -0,0 +1,438 @@
...
+   * - `7.11. The import statement <https://docs.python.org/3/reference/simple_stmts.html#the-import-statement>`_
+     - Not Supported
+     -
+   * - `7.11.1. Future statements <https://docs.python.org/3/reference/simple_stmts.html#future-statements>`_
+     - Not Supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I think we require this. we throw if you try to run ` 1 / 2 ` on old python without future statements

File: docs/source/jit_language_reference.rst [hunk 33]
```diff
@@ -0,0 +1,438 @@
...
+     -
+   * - `7.13. The nonlocal statement <https://docs.python.org/3/reference/simple_stmts.html#the-nonlocal-statement>`_
+     - Not Supported
+     -
+   * - `8. Compound statements <https://docs.python.org/3/reference/compound_stmts.html#>`_
+     - Irrelevant
+     -
+   * - `8.1. The if statement <https://docs.python.org/3/reference/compound_stmts.html#the-if-statement>`_
+     - Partially Supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I would just say supported, string parser not user-facing

File: docs/source/jit_language_reference.rst [hunk 34]
```diff
@@ -0,0 +1,438 @@
...
+     - String parser doesn't like semicolons
+   * - `8.2. The while statement <https://docs.python.org/3/reference/compound_stmts.html#the-while-statement>`_
+     - Partially Supported
+     - while..else doesn't work (and probably shouldn't), but we should fix the error
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Rephrase to be user-facing: `while ... else not supported`

File: docs/source/jit_language_reference.rst [hunk 35]
```diff
@@ -0,0 +1,438 @@
...
+   * - `8.3. The for statement <https://docs.python.org/3/reference/compound_stmts.html#the-for-statement>`_
+     - Partially Supported
+     - Document what can be iterated over, same issues as with while..else
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: I think we can just say supported. we have other sections for iteration.

File: docs/source/jit_language_reference.rst [hunk 36]
```diff
@@ -0,0 +1,438 @@
...
+   * - `8.4. The try statement <https://docs.python.org/3/reference/compound_stmts.html#the-try-statement>`_
+     - Not Supported
+     -
+   * - `8.5. The with statement <https://docs.python.org/3/reference/compound_stmts.html#the-with-statement>`_
+     - Not Supported
+     -
+   * - `8.6. Function definitions <https://docs.python.org/3/reference/compound_stmts.html#function-definitions>`_
+     - Not Supported
+     - These are kind of supported for symbolic_script, but closures cannot be called and cannot be created in the Python frontend
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Symbolic script isn't user facing. I would just say: "Nested function definitions (closures) not supported"

File: docs/source/jit_language_reference.rst [hunk 37]
```diff
@@ -0,0 +1,438 @@
...
+   * - `8.7. Class definitions <https://docs.python.org/3/reference/compound_stmts.html#class-definitions>`_
+     - Not Supported
+     - We need to make clear what is supported in Python land vs inside the TorchScript language
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: Rephrase: `We need to make clear what is supported in Python land vs inside the TorchScript language` to be user-facing. I would say something like, `nn.Modules` are supported, support for other classes is experimental

File: docs/source/jit_language_reference.rst [hunk 38]
```diff
@@ -0,0 +1,438 @@
...
+     - Not Supported
+     -
+   * - `8.8.3. The async with statement <https://docs.python.org/3/reference/compound_stmts.html#the-async-with-statement>`_
+     - Not Supported
+     -
+   * - `9. Top-level components <https://docs.python.org/3/reference/toplevel_components.html#>`_
+     - Not Relevant
+     -
+   * - `9.1. Complete Python programs <https://docs.python.org/3/reference/toplevel_components.html#complete-python-programs>`_
+     - Partially Supported
```
Comment (eellison) [inline docs/source/jit_language_reference.rst]: This seems not relevant to me. This section mostly links to the invocation from `__main__` which doesnt make sense for us.

**Review by eellison — COMMENTED** [review]
Looks really good! Collection of brain dump / comments. Don't feel the need to accept every comment

**Comment by eellison** [comment]
Also, it's going to be hard to not make this bit-rot like crazy. Any thoughts on how we'll keep it updated? For the jit unsupported stuff, i added tests. That doesn't have to be the only approach.

**Comment by driazati** [comment]
I don't think this page is going to get a lot of traffic, it's very in the weeds to a point that most people probably won't care about it. Its main use would be for comparisons to other similar projects to get a quick estimation of what is supported, so I don't think we should spent a ton of time keeping it up to date outside of syncing with the features in each release.

There is a point at which this is "done", i.e. when everything we're going to do it marked "Supported", so the bit rot really only goes in 1 direction (we fixed something and forgot to change it here)

**Comment by eellison** [comment]
So the answer is: we'll re-review it before every release and update it ? that seems fine to me. do you mind being an owner of this ?

**Comment by driazati** [comment]
Updated with a lot of the fixes you suggested (most of the formatting stuff was from the `m2r` conversion messing things up)

File: docs/source/jit_python_reference.rst [hunk 39]
```diff
@@ -0,0 +1,433 @@
...
+     -
+   * - `7.1. Expression statements <https://docs.python.org/3/reference/simple_stmts.html#expression-statements>`_
+     - Supported
+     -
+   * - `7.2. Assignment statements <https://docs.python.org/3/reference/simple_stmts.html#assignment-statements>`_
+     - Supported
+     -
+   * - `7.2.1. Augmented assignment statements <https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements>`_
+     - Partially Supported
+     - See delimiters section
```
Comment (eellison) [inline docs/source/jit_python_reference.rst]: ref would b nice

File: docs/source/jit_python_reference.rst [hunk 40]
```diff
@@ -0,0 +1,433 @@
...
+     - Python values cannot be captured
+   * - `4.3. Exceptions <https://docs.python.org/3/reference/executionmodel.html#exceptions>`_
+     - Partially Supported
+     - See ``try`` and ``raise`` statement section
+   * - `5. The import system <https://docs.python.org/3/reference/import.html>`_
+     - Not Relevant
+     -
+   * - `6. Expressions <https://docs.python.org/3/reference/expressions.html#>`_
+     - Not Relevant
+     - See expressions section
```
Comment (eellison) [inline docs/source/jit_python_reference.rst]: Ref would be nice

**Review by eellison — APPROVED** [review]
LGTM. Next time it would be nice to not force-push so i can look at incremental changes

File: docs/source/jit_python_reference.rst (hunk 39 above)
Comment (driazati) [reply docs/source/jit_python_reference.rst]: I don't think you can add references to other table rows, so we're stuck with this as is afaik
